### ISSUE TITLE
Crash when using mouse (int mode) and ne2000 networking

### ISSUE THREAD

**Issue by SuperIlu** [dosbox-staging/dosbox-staging#2180]
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### What Operating System are you using?

Windows 10

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

While developing [DOStodon](https://github.com/SuperIlu/DOStodon) I noticed erratic crashes.
I was able to track it down to the mouse support being enabled (`seamless` or `onstart`).
These crashes happen when the mouse is moved in/over the DOSBox window for a while.
This only happens with DOSBox-staging, DOSBox-X and real hardware seem to have no problem with the mouse.
DOStodon is based on [DOjS](https://github.com/SuperIlu/DOjS) which itself is based on [Allegro-XC](https://github.com/msikma/allegro-4.2.2-xc).
Stack analysis using debug builds of DOjS and SYMIFY.EXE suggest a problem with INT driven mouse updates.
I tried to reproduce with plain DOjS w/o networking without success.

This bug _might_ be related to https://github.com/dosbox-staging/dosbox-staging/issues/2056

![image](https://user-images.githubusercontent.com/1155988/210148827-cd64d66e-b453-45e5-8ef6-7337ac603485.png)

### Steps to reproduce the behaviour.

Explain how to reproduce
1. Install dosbox-staging-windows-msys2-x86_64-v0.80.0.zip
2. Install [ne2k driver](http://www.georgpotthast.de/sioux/pktdrv/ne2000.zip)
3. Use supplied config
4. Run [DOStodon](https://github.com/SuperIlu/DOStodon/archive/refs/heads/main.zip)
5. Move mouse in/out of DOSBox window quite often
6. Wait for crash to happen

[console-log.txt](https://github.com/dosbox-staging/dosbox-staging/files/10327603/console-log.txt)

### Your configuration

```ini
[dosbox]
memsize            = 32

[mouse]
mouse_capture        = seamless
#mouse_capture        = onstart

[ethernet]
ne2000            = true
nicbase           = 300
nicirq            = 3
macaddr           = AC:DE:48:88:99:AA

[autoexec]
# Lines in this section will be run at startup.
# You can put your MOUNT lines here.
mount c C:\Users\iluvatar\Desktop\DBS-mouse\drive_c
c:
ne2000 0x60 3 0x300
```

### Provide a Log

```text
console output is attached above
[console-log.txt](https://github.com/dosbox-staging/dosbox-staging/files/10327603/console-log.txt)
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

**Comment by FeralChild64** [comment]
@SuperIlu Do you have this problem also with CtMouse?

I have tried to reproduce the problem, but so far I haven't seen any crash, just the application getting stuck.

**Comment by kcgen** [comment]
I tried this reproduction configuration as well as the recommended (`nomouse`) configuration, but struggling to figure out what to do after the program loads:

https://user-images.githubusercontent.com/1557255/210152792-68f6b838-5bd3-463a-9786-49aebed82789.mp4

Surely PEBKAC on my side, help! :sweat_smile:

**Comment by SuperIlu** [comment]
I'd guess there is a problem connecting to the mastodon-server?
Can you provide the `JSLOG.TXT`? **Please make sure you wipe your credentials from it!**
I'd love to provide a more stripped down way to reproduce than a full flexed DOStodon/DOjS, but I didn't manage yet...

@FeralChild64 CtMouse in DOSBox? Is that possible?

**Comment by johnnovak** [comment]
Wouldn't using a sanitizer build help pinpoint the issue?

**Comment by FeralChild64** [comment]
@SuperIlu CtMouse 2.1 should work without any problems, it just replaces existing mouse driver. Unfortunately, not all PS/2 mouse driver work currently, as some need register-level access (not emulated - it’s a bit hard, unfortunately) and currently only BIOS part is provided. But you can also add a COM port mouse in [serial] secion.

Some mouse drivers refuse to load if they detect another mouse driver is already running - in such case you have to disable the built-in DOS driver n [mouse] section.

It’s quite useful for testing, can quickly limit the area to look for a problem.

**Comment by SuperIlu** [comment]
@FeralChild64 thank you. I'll try CtMouse when I have a chance.

**Comment by kcgen** [comment]
> I'd guess there is a problem connecting to the mastodon-server?
Can you provide the JSLOG.TXT? Please make sure you wipe your credentials from it!

Looks like it! Thanks for the help: [JSLOG.TXT.gz](https://github.com/dosbox-staging/dosbox-staging/files/10328026/JSLOG.TXT.gz)

The servername, username, and password I'm using does work when logging in via the web interface; however, maybe there's some syntax needed? 

Any UI improvements would be much appreciated, along with exact examples.

For example: should the server name be bare or with URI header?
 - `fosstodon.org` vs. `https://fosstodon.org` vs. `mastason://fostodon.org/PORT/...`

Should the username be the alias, account, account@host or alias@account@host? 
 - `alias` vs `account@fosstodon.org` vs `alias@account@fosstodon.org`

**Comment by kcgen** [comment]
> Wouldn't using a sanitizer build help pinpoint the issue?

Definitely couldn't hurt on our side to catch compounding problems like leaks (or things like that).

**Comment by johnnovak** [comment]
> > Wouldn't using a sanitizer build help pinpoint the issue?
> 
> Definitely couldn't hurt on our side to catch compounding problems like leaks (or things like that).

Yeah I was thinking @SuperIlu could use a sanitizer build (instructions can be found [here](https://github.com/dosbox-staging/dosbox-staging/blob/main/BUILD.md#make-a-sanitizer-build)) and send us his results, as @FeralChild64 seems to have problems with reproducing it.

**Comment by SuperIlu** [comment]
@kcgen damn, this is a known issue I still could not reproduce on my side:
```
Logging in...Error: Login failed: 400: {"error":"invalid_grant","error_description":"The provided authorization grant is invalid, expired, revoked, does not match the redirection URI used in the authorization request, or was issued to another client."}
```
I have test accounts on two different servers and I even asked in the Mastodon developer Discord an nobody could give me a clue what is going on...

Regarding your questions:
- server name: with or without `https://` is fine. DOStodon just needs domain name (`fosstodon.org` or `https://fosstodon.org` should both work)
- Username is the email-address shown on the "Account settings" page of your account.
- Password is the accounts password, 2FA is not supported, passwords with characters that are not 7bit ASCII may not work.

But you are right, I should update the documentation on the DOStodon GitHub.
And I also should quit the program with a meaningful error message if login fails I guess :)

Sorry for all the hassle...

**Comment by kcgen** [comment]
> Sorry for all the hassle...

No worries; DOSBox Staging has generated almost 800 hassles! :sweat_smile:  

![2022-12-31_15-17](https://user-images.githubusercontent.com/1557255/210159582-8d270e16-8b93-49f7-a44f-66ac296dbb88.png)

Thanks for clarifying the login details, @SuperIlu. I now see content.

I was able to reproduce the period crash, however with the IRQ set to 10, I haven't hit it yet:

```ini
[ethernet]
ne2000  = true
nicbase = 300
nicirq  = 10
macaddr = AC:DE:48:88:99:AA

[autoexec]
@echo off
mount c DOStodon-main
c:
ne2000 0x60 10 0x300
```

Curious if that helps?

**Comment by SuperIlu** [comment]
I'll try switching to irq10 and also using CtMouse on my end today.

And maybe you just helped me find the weird login-bug? Maybe people did use their Mastodon-Account name instead of the mailaddress?!?

**Comment by SuperIlu** [comment]
This is what I found:
- I can still reproduce the problem with IRQ10 for the NIC.
- The problem still happens when CTMOUSE.EXE (v2.1b4) is loaded on top of DOSBox mouse driver

**Comment by johnnovak** [comment]
Side comment: be aware of Cute Mouse; I just noticed in some games it creates some unwanted side effects.

Day of the Tentacle, CD version: with CTMOUSE.EXE (v2.1b4) mouse button release events would be registered as mouse down events in-game. Ugh...

All I'm saying, don't just trust it that it will always do the correct thing 😄 There were boatloads of mouse drivers back in the day for a reason...

**Comment by SuperIlu** [comment]
If I remember correctly I have a mainboard in my stash where v2.1 will not work, only an older version...

**Comment by FeralChild64** [comment]
But at least CtMouse allows to rule out some parts of DOSBox as a potential problem source.

**Comment by SuperIlu** [comment]
Update to the error description and new instructions to reproduce after I had some time to dig into this again:

I could nail down the issue to a minimal test case using [Allegro](https://github.com/msikma/allegro-4.2.2-xc) and the following code (stripped down from [exmouse.c](https://github.com/msikma/allegro-4.2.2-xc/blob/master/examples/exmouse.c):
```
#include <stdio.h>

#include <allegro.h>

int main(void) {
    int mickeyx = 0;
    int mickeyy = 0;
    BITMAP *custom_cursor;
    int c = 0;

    if (allegro_init() != 0) return 1;
    install_keyboard();
    install_timer();

    set_color_depth(32);
    if (set_gfx_mode(GFX_AUTODETECT, 640, 480, 0, 0) != 0) {
        if (set_gfx_mode(GFX_SAFE, 320, 200, 0, 0) != 0) {
            set_gfx_mode(GFX_TEXT, 0, 0, 0, 0);
            allegro_message("Unable to set any graphic mode\n%s\n", allegro_error);
            return 1;
        }
    }

    set_palette(desktop_palette);
    clear_to_color(screen, makecol(255, 255, 255));

    /* Detect mouse presence */
    if (install_mouse() < 0) {
        textout_centre_ex(screen, font, "No mouse detected, but you need one!", SCREEN_W / 2, SCREEN_H / 2, makecol(0, 0, 0), makecol(255, 255, 255));
        readkey();
        return 0;
    }

    poll_mouse();

    clear_to_color(screen, makecol(255, 255, 255));
    textprintf_centre_ex(screen, font, SCREEN_W / 2, 8, makecol(0, 0, 0), makecol(255, 255, 255), "Driver: %s", mouse_driver->name);
    show_mouse(screen);
    readkey();
    show_mouse(NULL);

    destroy_bitmap(custom_cursor);

    return 0;
}

END_OF_MAIN()
```

Used `dosbox-staging.conf` to reproduce:
```
[dosbox]
memsize            = 32

[mouse]
mouse_capture        = seamless

[autoexec]
mount c C:\Users\iluvatar\Desktop\DBS-mouse\drive_c
c:
```

![image](https://user-images.githubusercontent.com/1155988/215336127-7f0cfa21-c399-4540-8715-f15fd4b5dcf1.png)

- This testcase breaks with DOSBox-staging when the mouse cursor leaves the window.
- It works fine with FreeDOS and CtMouse 2.1 beta4 (where it can't leave the screen).
- The `SYMIFY.EXE` stacktrace points to `__dmpi_yield()`. If I comment out the call in [sys_dos_yield_timeslice()](https://github.com/msikma/allegro-4.2.2-xc/blob/master/src/dos/dsystem.c#L621) it does not crash. 
- The crash does also not occur when `show_mouse(screen);` is commented out in the example.
- For each testrun Allegro autoselected the `Mickey mouse`-driver from [dmouse.c](https://github.com/msikma/allegro-4.2.2-xc/blob/master/src/dos/dmouse.c#L84)
- Despite my previous observations, this crash does also occur without ne2000 support

I attached the compiled [binaries.zip](https://github.com/dosbox-staging/dosbox-staging/files/10530094/binaries.zip) with and without `__dmpi_yield()`.

**Comment by joncampbell123** [comment]
Simple question: If you disable the emulator INT 33h, enable AUX or serial mouse emulation and load a mouse driver TSR instead, does it crash?

**Comment by joncampbell123** [comment]
Actually __dpmi_yield() makes a specific INT 2Fh call. https://www.delorie.com/djgpp/doc/libc/libc_279.html

First check: Can you see if the INT 2Fh real mode vector is non-NULL (not 0000:0000)?

Second check: Is there anything in the native code INT 2Fh callback handler, starting from INT2F_Handler in src/dos/misc.cpp, that might access memory at the wrong time when the CPU is possibly in protected mode instead of virtual 8086 mode?

**Comment by joncampbell123** [comment]
One function to look at might be DOS_MultiplexFunctions which has the majority of INT 2Fh handling code. Anything else that calls DOS_AddMultiplexHandler should be examined as well.

**Comment by SuperIlu** [comment]
Regarding the questions from @joncampbell123 I feel I'm able to answer:
- I tested with CtMouse 1.9 and 2.1b4
- The crash does happen with an identical stack trace no matter if I use the builtin and any of the CtMouse drivers
- It does not matter if the mouse is "connected" to PS/2 or COM1

**Comment by kcgen** [comment]
@SuperIlu , is it possible to log the mouse coordinates allegro receives?  it would be interesting to see if (or how) they differ when running ctmouse vs. without ctmouse.

Maybe ctmouse is bounding or limiting the values just within allegro's ability to handle them, while dosbox's coordinates are allowed to go "a bit too far" (which Allegro currently isn't handling properly).

**Comment by SuperIlu** [comment]
sorry if that was unclear: using CtMouse does not change anything, the crash happens as well with both versions of CtMouse...

**Comment by kcgen** [comment]
Tagging issue w/ help-wanted.

Thanks for clarifying, @SuperIlu.

**Comment by joncampbell123** [comment]
Testing with DOSBox-X on Mac OS, m_broken doesn't appear to segfault at all unless I move the cursor to any edge of the screen, but it takes a fair amount of time for that to happen.

m_works doesn't seem to exhibit the issue.

**Comment by joncampbell123** [comment]
I then tried attaching a Logitch mouse (the gamer type) that I know has a high report rate and tried again. m_works doesn't have any issues, m_broken only segfaults if I move the mouse cursor along any edge of the screen.

**Comment by joncampbell123** [comment]
Which leads me to another question: How much stack does your program allocate for itself? Perhaps there is something with Allegro that uses just enough stack that the additional stack needed for the DPMI call and/or the INT 33 callback might overflow it and cause a crash. See if you can make those samples again instructing the linker to provide a bigger stack.

**Comment by SuperIlu** [comment]
I can confirm that the issue only happens when the mouse leaves the DOSBox window. For me it happens immediately with m_broken.

Regarding the stack: I checked with [DJGPPs FAQ](http://www.delorie.com/djgpp/v2faq/faq15_9.html) some time ago for another issue and it states that there is 512KiB of stack available as a default.
I can try increasing that, but I think it is already plenty...

**Comment by joncampbell123** [comment]
I just had another thought, because this might not be obvious if you're used to modern OSes:

Does your INT 33h mouse handler reside in locked memory? I mean page locked with the DPMI server.

Some DPMI servers like CWSDPMI automatically make a swap file on disk. It can manage a swap file just fine, the problem is what happens if your interrupt handler is paged out to disk. MS-DOS is not re-entrant, so if a page fault happens from an interrupt handler, the DOS extender has no choice but to try to bail out to DOS and report a segfault.

I also wonder if the server might page out anything from conventional memory, which seems unlikely as far as I know, but perhaps all it takes is an INT 2Fh to trigger that.

Perhaps locking your stack in memory with the DPMI server might help as well.

Can you try running your m_broken example with a DOS extender and DPMI server that *does not* use paging or a swap file of any kind?

**Comment by SuperIlu** [comment]
> I just had another thought, because this might not be obvious if you're used to modern OSes:
> 
> Does your INT 33h mouse handler reside in locked memory? I mean page locked with the DPMI server.

I have no clue what Allegro does there. 

> Some DPMI servers like CWSDPMI automatically make a swap file on disk. It can manage a swap file just fine, the problem is what happens if your interrupt handler is paged out to disk. MS-DOS is not re-entrant, so if a page fault happens from an interrupt handler, the DOS extender has no choice but to try to bail out to DOS and report a segfault.

DOjS uses the CWSDPMI-Version with swapping disabled, I need to check which one I used with the test-cases.
Also: I can't see why it would swap out the above program on a box with 32MiB of memory?


> Perhaps locking your stack in memory with the DPMI server might help as well.

I have no clue how to do that with Allegro/DJGPP.

> Can you try running your m_broken example with a DOS extender and DPMI server that _does not_ use paging or a swap file of any kind? DOS4/GW and DOS32A, if run without EMM386.EXE loaded, are known to load your code into extended memory and run in a flat 32-bit segment without any paging.

AFAIK DJGPP EXEs load CWSDPMI by default, how can I change that?

**Comment by ThomasEricB** [comment]
> This bug _might_ be related to #2056
@SuperIlu You are correct. This is the same issue I reported at #2056 but outside of Windows 9x. 

Hopefully @FeralChild64 's recent pull https://github.com/dosbox-staging/dosbox-staging/pull/2287 fixes this issues. 

Tagging @johnnovak and @kcgen for visibility

**Comment by FeralChild64** [comment]
IMHO it is unlikely that my PR fixes this particular issue.

**Comment by ThomasEricB** [comment]
> IMHO it is unlikely that my PR fixes this particular issue.

Don't be a pessimist. Even if it doesn't, it's still a step forward towards fixing this issue! Besides I been seeking Microsoft IntelliMouse emulation for Staging since I ran Windows 98 for the first time in it!

**Comment by FeralChild64** [comment]
This is probably a long standing issue, not necessary linked to NE2000 - see https://github.com/joncampbell123/dosbox-x/issues/3262; both DOSBox-X and DOSBox SVN have problems with Allegro too.

**Comment by SuperIlu** [comment]
yes, see my update here: https://github.com/dosbox-staging/dosbox-staging/issues/2180#issuecomment-1407695420

**Comment by FeralChild64** [comment]
I have reproduced the problem using https://github.com/joncampbell123/dosbox-x/files/8190790/test-file.zip (test file from DOSBox-X) - can be reproduced easily using built-in mouse driver, CtMouse, Microsoft Mouse driver 8.20, or Mouse Systems Mouse driver 8.00, A4 Tech mouse driver 8.04A, DR-DOS mouse drvier R1.1  - doesn't matter.

I have created a debug build, and set `core = normal` in the configuration file. Result - this time emulator crashes, no matter which mouse driver I use:

```
Stack trace:
21      0x558eaf5e2bd1 ./build/dosbox(+0x4bbd1) [0x558eaf5e2bd1]
20      0x7f734a64e4c5 __libc_start_main + 133
19      0x7f734a64e40a /lib64/libc.so.6(+0x2340a) [0x7f734a64e40a]
18      0x558eaf5e2cc2 main + 32
(...)
2023-06-29 19:57:32.437 | ABORT: Pagefault didn't correct page
Aborted (core dumped)
```

This might be a CPU emulation bug after all, not necessary a mouse related problem.

**Comment by johnnovak** [comment]
Can we move this out of the v0.81.0 scope @FeralChild64 ?

**Comment by kneelian** [comment]
The bug still replicates easily with 0.81.2. I am using the `test.exe` Allegro binary as distributed via DJGPP's zip picker, which crashes when the mouse pointer hits one of the boundary pixels.

I used to be able to replicate this bug with a previous version of DOSBox-X. Updating to DOSBox-X version 2024.07.01 (the stable one as it stands today) seems to solve the problem, as I've been mashing the mouse around the window for a bit without any GPFs; I don't have a good way of figuring out at which point it got solved (perhaps a bisect would work here?), but it's a good starting point.

Below is an example GPF as it stands on DOSBox-Staging v0.81.2. Note that the trace is inconsistent, so this is just one of the examples I get.

![image](https://github.com/user-attachments/assets/dead63cc-7157-4fef-b5d8-cdd68f1aa25a)

**Comment by spagbolninja** [comment]
I have been running into the same issue. I have installed DOSBox-X v2024.10.01 on Debian using flatpak after seeing the above comment (I had originally encountered this issue on Staging.)

I have been using it to play around with the old Allegro 4.2.3 libraries and DJGPP. This is the version from 'Mr.Dictionary's' site which fixes a bug in Allegro's sound blaster libraries.   (https://www.mrdictionary.net/allegro/) 

When using the Allegro test/sample programs I can reproduce the same error when the mouse cursor touches the edge of the window (or screen if I change the config file to start in full screen mode.) 

Modifying DosBox-X's config file and: Locking aspect ratio, turning off mouse emulation and lowering mouse sensitivity, have had no effect on this reproducable behavior. 

However, once the error has occurred once, if I restart allegro programs from the command line of the same open instance of DosBox-X, I can then move the mouse to the corner of the screen without issue. Whatever is causing this only occurs that first time (until DosBox-X is closed and opened again.)

I had previously encountered this error on DosBoX staging and saw the above post and decided to give DosBox-X a try. While it hasn't cured the issue, **Staging will always crash when the mouse touches a window/screen border whereas, as mentioned above, X will only do it that first time.**

Hope any of this information might help!

**Comment by johnnovak** [comment]
@spagbolninja Please raise a separate issue ticket with exact repro steps and attach the test program binary as well, then link to this ticket. Thanks.

**Comment by SuperIlu** [comment]
> I have been using it to play around with the old Allegro 4.2.3 libraries and DJGPP. This is the version from 'Mr.Dictionary's' site which fixes a bug in Allegro's sound blaster libraries. (https://www.mrdictionary.net/allegro/)

If you compile Allegro on your own please see my comment about `sys_dos_yield_timeslice()` here https://github.com/dosbox-staging/dosbox-staging/issues/2180#issuecomment-1407695420

If you comment out `__dpmi_yield()` you can run Allegro programs with mouse in DOSBox-X and DOSBox-staging. I did that in DOjS, see https://github.com/SuperIlu/DOjS/blob/cac87b8c21af68c884b63a57bb62b0876515f51b/3rdparty/allegro-4.2.2-xc-master/src/dos/dsystem.c#L624

**Comment by SuperIlu** [comment]
> Thank you for the heads up! Unfortunately when I commented out __dpmi_yield() in /allegro/src/dos/dsystem.c and recompiled Allegro's Setup (Allegro/Setup/Setup.c) and Test (Allegro/Tests/Test.c) (https://github.com/spagbolninja/testrepository/blob/main/TEST.C) using DJGPP within Dosbox-Staging the issue remains for me where it crashes with a SIGSEGV error.

Sorry if I misunderstand, but I'll ask anyway just to be sure:
Did you recompile the while liballegro and then build Setup.c and Test.c? Or did you just recompile the two examples?
For the fix to work you need to rebuild the whole library.

Switching between the different DOS mouse drivers didn't change the behavior for me, commenting out `__dpmi_yield()` did.
Can you check if the binaries I attached to https://github.com/dosbox-staging/dosbox-staging/issues/2180#issuecomment-1407695420 work on your setup?
one should crash, the other should work...

**Comment by spagbolninja** [comment]
> Sorry if I misunderstand, but I'll ask anyway just to be sure: Did you recompile the while liballegro and then build Setup.c and Test.c? Or did you just recompile the two examples? For the fix to work you need to rebuild the whole library.

Ah thank you! My mistake! I can confirm that this worked!

- I recompiled Allegro from within Staging using DJGPP, after commenting out **__dmpi_yield** in allegro/src/dos/dsystem.c as you noted
- 'make installall' encountered an error with 'makedoc.exe' and 'ahack.txt', (however this seems to have been right at the end of the process and hasn't affected functionality
- 'gcc test.c -o test.exe' now compiles an executable where the mouse-touching-border crash has not occured

I can also confirm that your 'working' binary file does indeed work and the mouse can touch the screen borders with no issues. Thanks!

**Comment by Grounded0** [comment]
Our current understanding is that this behaviour is caused by this:

- #4461

Allegro library requires CPU page fault handling from the system:

- #3987

@johnnovak 

Could you modify labels for this to include only bug + CPU/FPU emulation?

**Comment by johnnovak** [comment]
@Grounded0 Good points, done.
